requests-for-comments/2023-11-22-scheduled-flag-changes.md

```diff
@@ -0,0 +1,84 @@
+# Request for comments: Scheduled Feature Flag changes
+
+## Problem
+
+Currently, users can make updates to feature flags only manually and in real-time. This is not ideal, as it potentially leads to overlooked or forgotten changes, particularly when they fall outside normal working hours. This can lead to poorly timed releases, negatively impacting end users' engagement and feedback.
+
+## Solution
+
+Scheduling feature flag changes allows users to automatically change flag properties at a future point in time. For example, a user can schedule an initial release of the flag for internal testing, then enable it to the beta users group 2 days later, and finally roll it out to 100% of the users in 2 weeks.
+
+## Scope
+
+This feature should allow the following scheduled modifications to a feature flag:
+
+-   **Enable/disable the flag** (field: `is_active`)
+-   **Add rollout conditions to the flag** (field: `filters`)
+
+We want to be able to _generalize_ this implementation to other models such as `Survey` or `Experiment`.
+
+## Implementation
+
+To schedule future changes, we can have a scheduling API, specifying the model the operate on, the id, the changes to be executed, and the schedule time.
+
+```
+POST /api/projects/:project_id/schedule
+{
+    "model": "FeatureFlag",
+    "id": 57,
+    "changes": [ ... ]
+    "scheduled_at": <timestamp>
+}
+```
+
+A change to the record can be expressed as a JSON, specifying the field to be modified and the new value to be set:
+
+**Example 1: Enable the flag**
+
+```
+{
+    "field": "is_active",
+    "value": "false"
+}
+```
+
+**Example 2: Add a rollout condition to the flag**
+
+```
+{
+    "field": “filter”,
+    "value": {
+        "variant": null,
+        "properties": [{
+            "key": "$geoip_country_name",
+            "type": "person",
+            "value": ["Australia"],
+            "operator": "exact"
+        }],
+        "rollout_percentage": 20
+    }
+}
+```
+
+_At_ or _after_ the time specified, the change needs to be propagated onto the record.
+
+We already use Celery with a Redis queue to execute various background tasks. However, because the Redis queue is not durable, we cannot schedule tasks at the precise point in the future. A possible solution here is to have Postgres table of scheduled changes. Celery could periodically poll this table, e.g. every minute. During each poll, Celery would process any rows where the _scheduled_at_ time has already passed.
+
+The `scheduled_flag_changes` table would have the following schema:
+
+| team_id | id  | changes (JSON) | scheduled_at (timestamp) | status (enum) |
```

Comment:
If you plan on having only two states for `status` (waiting, done), I'd recommend swapping it for an `executed_at` nullable column. This way, you can compute statistics about how many tasks have been executed within proper timing. That would be your [SLI](https://www.atlassian.com/incident-management/kpis/sla-vs-slo-vs-sli) for that feature.

You'll want to remove old rows eventually. What retention would be OK for customers?

Reply:
- Makes sense. Initially, I thought about three states (waiting, claimed, done), but since the operation is idempotent, two states will suffice. No need to claim the row.

- Yes, good point; I'd set it arbitrarily to 90 days and tweak it later as needed.

Reply:
> No need to claim the row (and having to figure out what happens when the worker that claimed the row fails).

Indeed, no need to claim it if concurrency = 1, then for higher concurrency I recommend using a `select for update skip locked` approach that does not mutate the row, using transactions instead.

Reply:
oh nice, I didn't know there's an easy way to force concurrency with celery, I assumed given 6 pods running celery workers, all talking to the same db, we'll by default have higher concurrency, but if celery can manage this itself, works 👍

Reply:
Okay, is my understanding correct that we want to have concurrency = 1 for the current/foreseeable load?

Reply:
> oh nice, I didn't know there's an easy way to force concurrency with celery

I don't have much experience with Celery, so I am probably missing a big caveat here. Can you please check that? It might require using a dedicated task queue (I think it would be beneficial to split our big celery task corpus into separate queue executed by the same workers).

If that's not trivial, we can solve concurrency at the PG level when implementing the task.

Reply:
.. looking at celery docs, I'm not sure where this is configured: https://docs.celeryq.dev/en/latest/userguide/workers.html#concurrency, our current settings seem to suggest we are using the defaults, which means # of cores workers, multiplied by 6 for the different pods running the workers.

I don't think we'll change the defaults, as that affects all tasks, but is there a way to enforce concurrency for this specific task?

If not, I think we are already in the `select for update skip locked` approach land

Reply:
If we can use the `skip locked` query here, I don't see any benefit in limiting concurrency to 1. It will be more scalable to have concurrency = N from the start.

Reply:
with concurrency you'll need to make sure you don't run concurrent mutations to the same flag, and error handing will get a bit trickier. I think we're talking about implementation details while I don't know your design constraints:

- what delay in mutations is acceptable
- can mutation fails? do they need to be retried?
- do you need to guard against concurrent mutations on the same feature flag (could happen after a long celery incident, this might mess the final state).

I need to focus on another topic right now, but I suggest you think more about these constraints, then we can schedule some sync time to discuss the implementation details. If you want retries and concurrency, temporal might be a simpler way forward than rewriting our own queue implementation.

Reply:
great points. As a side note, I think we can still enforce concurrency=1 at db instead of celery level and skip the above problems by not adding 'skip locked' to the select, and wrapping it in a transaction. It's another option to consider if solving the above becomes too unwieldy.